File: src/main/java/cope/nebula/client/feature/module/world/FastPlace.java
package cope.nebula.client.feature.module.world;

import cope.nebula.asm.mixins.IMinecraft;
import cope.nebula.client.feature.module.Module;
import cope.nebula.client.feature.module.ModuleCategory;
import cope.nebula.client.value.Value;
import cope.nebula.util.world.entity.player.inventory.InventoryUtil;
import net.minecraft.init.Items;
import net.minecraft.item.Item;
import net.minecraft.item.ItemBlock;
import net.minecraft.util.EnumHand;

public class FastPlace extends Module {
    public FastPlace() {
        super("FastPlace", ModuleCategory.WORLD, "Places items faster");
    }

    public static final Value<Integer> speed = new Value<>("Speed", 4, 0, 4);

    public static final Value<Boolean> xp = new Value<>("XP", true);
    public static final Value<Boolean> crystals = new Value<>("Crystals", false);
    public static final Value<Boolean> blocks = new Value<>("Blocks", true);

    @Override
    public void onTick() {
        Item item = InventoryUtil.getHeld(EnumHand.MAIN_HAND).getItem();
        if (xp.getValue() && item.equals(Items.EXPERIENCE_BOTTLE) ||
                crystals.getValue() && item.equals(Items.END_CRYSTAL) ||
                blocks.getValue() && item instanceof ItemBlock) {

            ((IMinecraft) mc).setRightClickDelayTimer(4 - speed.getValue());
        }
    }
}

File: src/main/kotlin/wtf/nebula/client/feature/module/world/FastPlace.kt
package wtf.nebula.client.feature.module.world

import me.bush.eventbuskotlin.EventListener
import me.bush.eventbuskotlin.listener
import net.minecraft.init.Items
import net.minecraft.item.ItemBlock
import wtf.nebula.client.event.tick.TickEvent
import wtf.nebula.client.feature.module.Module
import wtf.nebula.client.feature.module.ModuleCategory

class FastPlace : Module(ModuleCategory.WORLD, "Places things fast and shit") {
    val speed by int("Speed", 4, 1..4)
    val ghostFix by bool("Ghost Fix", true)

    val blocks by bool("Blocks", true)
    val exp by bool("EXP", true)
    val fireworks by bool("Fireworks", true)
    val crystals by bool("Crystals", true)

    @EventListener
    private val tickListener = listener<TickEvent> {
        val holding = mc.player.heldItemMainhand.item
        if (blocks && holding is ItemBlock
            || exp && holding == Items.EXPERIENCE_BOTTLE
            || fireworks && holding == Items.FIREWORKS
            || crystals && holding == Items.END_CRYSTAL) {

            mc.rightClickDelayTimer = 4 - speed
        }

        if (ghostFix) {
            // TODO
        }
    }
}